routes/cube/helper.js

```diff
@@ -74,58 +74,58 @@ async function updateCubeAndBlog(req, res, cube, changelog, added, missing) {
 }
 
 async function bulkUpload(req, res, list, cube) {
-  const cards = list.match(/[^\r\n]+/g);
-  let missing = '';
+  const lines = list.match(/[^\r\n]+/g);
+  let missing = [];
   const added = [];
   let changelog = '';
-  if (cards) {
-    if ((cards[0].match(/,/g) || []).length > 3) {
-      let newCards = [];
-      let newMaybe = [];
+  if (lines) {
+    if ((lines[0].match(/,/g) || []).length > 3) {
+      // upload is in CSV format
+      let newCards;
+      let newMaybe;
       ({ newCards, newMaybe, missing } = CSVtoCards(list, carddb));
       changelog = newCards.reduce((changes, card) => changes + addCardHtml(carddb.cardFromId(card.cardID)), changelog);
       cube.cards.push(...newCards);
       cube.maybe.push(...newMaybe);
       added.concat(newCards, newMaybe);
     } else {
-      for (const itemUntrimmed of cards) {
+      // upload is in TXT format
+      for (const itemUntrimmed of lines) {
         const item = itemUntrimmed.trim();
-        const numericMatch = item.match(/([0-9]+)x? (.*)/);
-        if (numericMatch) {
-          let count = parseInt(numericMatch[1], 10);
-          if (!Number.isInteger(count)) {
-            count = 1;
-          }
-          for (let j = 0; j < count; j += 1) {
-            cards.push(numericMatch[2]);
+        // separate counts and sets from the name
+        //                              |    count?   |name|     (set)?      |
+        const splitLine = item.match(/^(?:([0-9]+)x? )?(.*?)(?: \(([^)(]+)\))?$/);
+        const name = splitLine[2];
+        const set = splitLine[3];
+        let count = parseInt(splitLine[1], 10);
```

Comment:
There's toNullableInt you could use here so it's just `toNullableInt(splitLine[1]) ?? 1`.

Reply:
can't use `??` yet.